File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.java
class Solution {
     public void solve(int m, int n , int ar[][]){
        //UP
        for(int i=m-1;i>=0;i--){
            if(ar[i][n] ==2 || ar[i][n]==3)break;
            ar[i][n]=1;
        }
        //DOWN
        for(int i=m+1;i<ar.length;i++){
            if(ar[i][n] ==2 || ar[i][n]==3)break;
            ar[i][n]=1;
        }
        //LEFT
        for(int i=n-1;i>=0;i--){
             if(ar[m][i] ==2 || ar[m][i]==3)break;
            ar[m][i]=1;
        }
         //RIGHT
        for(int i=n+1;i<ar[0].length;i++){
             if(ar[m][i] ==2 || ar[m][i]==3)break;
            ar[m][i]=1;
        }

     }
    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {
        int ar[][]= new int[m][n];

        for(int i[]:ar)
        Arrays.fill(i,0);

        for(int i[]:guards) ar[i[0]][i[1]]=2;
        
        for(int i[]:walls) ar[i[0]][i[1]]=3;
        
        for(int i[]:guards)
        solve(i[0],i[1],ar);

        int c=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(ar[i][j]==0)c++;
            }
        }
        return c;

    }
}

File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.cpp
class Solution {
public:
    void dfs(int i, int j, int m, int n, int& nVisited, vector<vector<int>>& vis, int dirX, int dirY){
        if(i<0 || j<0 || i>=m || j>=n || vis[i][j]==-1 || vis[i][j]==2) return;
        if(vis[i][j]==0) nVisited++;
        vis[i][j]=1;
        dfs(i+dirX, j+dirY, m, n, nVisited, vis, dirX, dirY);
    }
    int countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        vector<vector<int>> vis(m, vector<int>(n, 0));
        int nVisited=0;
        for(auto wall: walls) vis[wall[0]][wall[1]]=-1;
        for(auto guard: guards) vis[guard[0]][guard[1]]=2;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(vis[i][j]==2) {
                    dfs(i+1, j, m, n, nVisited, vis, 1, 0);
                    dfs(i-1, j, m, n, nVisited, vis, -1, 0);
                    dfs(i, j-1, m, n, nVisited, vis, 0, -1);
                    dfs(i, j+1, m, n, nVisited, vis, 0, 1);
                }
            }
        }
        return m*n - nVisited - walls.size() -guards.size();
    }
};